Improve progress bar displayed during ajax call

seballot opened the pull request:
**Why?**
Sometimes the progress bar pops in and out like a strobe

**What**
Improve style, use fontawesome spinner
Display it only when the request takes long time to complete (500ms)

![progress](https://user-images.githubusercontent.com/17404254/143575578-10155ed0-7f49-4a45-8a36-de31e280e4c7.gif)

*How to Test*
Look for a user with a slow network connexion

jibees reviewed (approved):
You're on fire @seballot ! Thanks a lot, very appreciate.
LGTM, but I re-trigger a build since some of the specs failed (to see if it's flaky spec or relevant)

seballot commented:
it's just small commits, easy to do !

Re specs, I tried to run them locally but couln't make the system test working ```Browser did not produce websocket url within 20 seconds, try to increase `:process_timeout`. See https://github.com/rubycdp/ferrum#customization```
any idea why? I'm on Debian, using docker

jibees commented:
Re specs.
Build is ok here. 🎉 
I don't use docker and I've no idea on how I can help you about that... All the core developers here at OFN don't use Docker anymore, so, it could be, unfortunately, mis-configured or with tech debt...

seballot commented:
ok no prob, will investigate if someday I need to fix or write systems specs

Matt-Yorkley reviewed (approved):
Thanks @seballot, this looks great :heart:

I just noticed there's a few lines of code that seem to duplicate this element here: https://github.com/openfoodfoundation/openfoodnetwork/blob/7ef976a52abbd5f189f76cd1d237dc27bf4c4522/app/views/spree/admin/taxonomies/edit.haml#L23-L26

I think we can probably just delete those lines (it just shows the same thing twice). It might be a nice bit of relevant tidying-up.

filipefurtad0 commented:
Thanks for this PR @seballot !

But, strange - this does not seem to have triggered a build in Semaphore, which we use manage deployments on staging servers:

![image](https://user-images.githubusercontent.com/49817236/144419101-3022e24e-bad6-41fe-b18a-6aa8ac9b145f.png)

Any idea why? Maybe other @openfoodfoundation/core-devs  can help? Added the feedback label.

Matt-Yorkley commented:
> But, strange - this does not seem to have triggered a build in Semaphore, which we use manage deployments on staging servers

Strange indeed! It won't let me manually build the branch either. I'll deploy it manually for now :ok_hand:

filipefurtad0 commented:
Ok, thanks for staging Matt :+1: 

Before this PR, we had indeed a strob-like and somewhat inconsistent behavior on the spinner:

![Peek 2021-12-02 11-56](https://user-images.githubusercontent.com/49817236/144436222-2dafc382-eaa7-44d1-9569-13d76d4a7603.gif)

After your PR @seballot it can be seen the spinner only appears during the (short, but connection dependent) time during which the results are being fetched:

![Peek 2021-12-02 13-04](https://user-images.githubusercontent.com/49817236/144427821-d02b4983-db98-49f5-87be-bea141adf603.gif)

This is awesome :tada: 

One remark though: we can see that the banner now is displayed in a different (orange) color than before (blue). After discussion with @openfoodfoundation/train-drivers-product-owners we've figured that the orange color could signal a warning or error to the users. Since the backoffice will go through some changes in the near future we'd try to introduce only the changes which are really needed right now.

Do you think it would be possible to add this small change to this PR, and bring back the blue color to the spinner? Other than that this looks great. Moving to In dev for now :+1: 

Many thanks for your initiative! :heart:

seballot commented:
@filipefurtad0 thanks for the test !
I've changed it to blue and remove the fixed width so it displays correctly in any languages (like in french in your screenshot)

![image](https://user-images.githubusercontent.com/17404254/145200275-653c6eed-9515-4a2f-b734-b80108058c83.png)


I also adjusted the loading messages so they look all the same in the admin pages

![image](https://user-images.githubusercontent.com/17404254/145199275-892a4049-1db4-4f45-bbeb-7436ec14a77d.png)

I did not create a reusable component because Angular directive are used differently, so I just left all templates use kind of duplicate code like
```haml
.sixteen.columns.alpha#loading{ 'ng-show' => 'RequestMonitor.loading' }
  = render partial: "components/spinner"
  %h1
    = t('.loading')
```

jibees commented on app/views/components/_spinner.html.haml:
```diff
@@ -1 +1 @@
-%img.spinner{ src: image_path("spinning-circles.svg"), style: "max-width: 100%" }
+%img.spinner{ src: image_path("spinning-circles.svg"), style: "max-width: 100%" }
```
I guess the only modification here is that the newline at the end of the file has been deleted. 
We used to keep those new lines on each file. I'd vote to put new line on this file too ;)

jibees commented on app/views/components/_admin_spinner.haml:
```diff
@@ -0,0 +1 @@
+%i.spinner.fa.fa-spin.fa-circle-o-notch
```
I'd add a new line too. ;)

jibees reviewed (approved):
Few comments, but looks good to me.

seballot commented:
Done :)

jibees commented:
@seballot 
This PR needs to be rebased... 🙏

seballot commented:
Hi @jibees, I did the rebase, but I wasn't able to test it cause the whole app is broken on my local machine due to changes on webpack (I'm using docker, I think I'm the only one)

Could you explain quickly what changed recently with webpack? so I can maybe fix the docker image?
I think it's just the webpack-dev-server that do no longer run correctly on docker

jibees commented:
@seballot Thanks, and sorry to hear that you had some issues around docker ; you're right, we don't use much Docker here. 

I know that @pacodelaluna opened a PR which is in `Dev Test Ready` state: https://github.com/openfoodfoundation/openfoodnetwork/pull/8518

Maybe it's a clue?

mkllnk reviewed (approved):
Thank you! There are some nice improvements in here.

I think that this problem could be solved better. The spinner was shown when Ajax is active. So if the spinner is flickering, we also have "flickering" Ajax requests which is bad for performance. It would be better to delay the request instead of delaying the spinner display.

Your solution is still an improvement though and I'm happy to approve it. :+1:

seballot commented:
Thanks @mkllnk, there is also a delay (debounce) with the ajax query made by select2, but when I tested it it was not working, maybe because we have an old version of select2.. I did not dig it too much
See ya

filipefurtad0 commented:
Hey @seballot ,

Hum, the spinner seems to be gone now:

![Peek 2022-01-24 19-45](https://user-images.githubusercontent.com/49817236/150853981-e5d76de5-6151-4ad5-bb5b-b4d2c2c50af3.gif)

Tested in Firefox, Chrome, Safari - with French and English locales: No spinner... Any clues?

filipefurtad0 commented:
Thanks for this tip @jibees :heart: 
(as of today on delivery circle)

We can see that the banner only appears on slow connections, simulate by enabling throttling on Chrome (slow 3G):
![Peek 2022-02-01 09-55](https://user-images.githubusercontent.com/49817236/151947964-df4ecc95-ea72-4fd4-b9f8-bbd9b97cf3ad.gif)

Translations also work:
![Peek 2022-02-01 09-57](https://user-images.githubusercontent.com/49817236/151948338-84c0db68-f7ab-4162-924c-02cd62e0f6d9.gif)

Merging, thanks @seballot
